File: init_firebase.js
const fs = require('fs');
const path = require('path');
const https = require('https');

const CONFIG_PATH = path.join(__dirname, 'firebase-config.js');
const DB_PATH = path.join(__dirname, 'db.json');

console.log("--------------------------------------------------");
console.log("🔥 Firebase Database Initializer & Seeding Script");
console.log("--------------------------------------------------\n");

// 1. Read firebase-config.js
if (!fs.existsSync(CONFIG_PATH)) {
  console.error("❌ Error: firebase-config.js not found! Please create it first.");
  process.exit(1);
}

const configContent = fs.readFileSync(CONFIG_PATH, 'utf8');

// Simple regex to extract databaseURL
const dbUrlMatch = configContent.match(/databaseURL\s*:\s*["']([^"']+)["']/);
if (!dbUrlMatch) {
  console.error("❌ Error: Could not find databaseURL in firebase-config.js.");
  process.exit(1);
}

const databaseURL = dbUrlMatch[1];
if (databaseURL.includes("YOUR_PROJECT_ID")) {
  console.error("❌ Error: firebase-config.js is still using placeholder values.");
  console.error("Please edit firebase-config.js with your actual Firebase Web App credentials first.");
  process.exit(1);
}

// 2. Read local db.json
if (!fs.existsSync(DB_PATH)) {
  console.error("❌ Error: Local db.json not found in this folder.");
  process.exit(1);
}

const dbContent = fs.readFileSync(DB_PATH, 'utf8');
const dbData = JSON.parse(dbContent);

// Transform data to support slashes and match Firebase expectations
const transformedData = {};

// 1. Settings
transformedData.settings = dbData.settings || {
  writeUp: "Welcome to Adira Telecom Allocation Portal. Note: Confirmed allocations are strictly locked and credit will only be extended post proof verification.",
  confirmationDeadline: "",
  paymentDeadline: ""
};

// 2. Prices - convert from key-value object to array of { modelName, price } objects
// to bypass Firebase Realtime Database key constraints on character "/"
transformedData.prices = [];
if (dbData.prices) {
  for (const [modelName, price] of Object.entries(dbData.prices)) {
    transformedData.prices.push({
      modelName: modelName,
      price: price
    });
  }
}

// 3. Allocations - keyed by ID
transformedData.allocations = {};
if (Array.isArray(dbData.allocations)) {
  dbData.allocations.forEach(item => {
    if (item.ID) {
      transformedData.allocations[item.ID] = item;
    }
  });
}

// 4. Distributors - keyed by escaped AD Name
transformedData.distributors = {};
if (Array.isArray(dbData.distributors)) {
  dbData.distributors.forEach(item => {
    if (item["AD Name"]) {
      const key = item["AD Name"].replace(/[.#$\[\]]/g, "_");
      transformedData.distributors[key] = item;
    }
  });
}

// 5. Users - keyed by Username
transformedData.users = {};
if (Array.isArray(dbData.users)) {
  dbData.users.forEach(item => {
    if (item.Username) {
      transformedData.users[item.Username] = item;
    }
  });
}

const payload = JSON.stringify(transformedData, null, 2);

console.log(`📡 Targeting Firebase Database: ${databaseURL}`);
console.log(`📦 Seeding settings, users (${Object.keys(transformedData.users).length}), allocations (${Object.keys(transformedData.allocations).length}), distributors (${Object.keys(transformedData.distributors).length})...`);

// 3. Make PUT request via Firebase REST API
const urlObj = new URL(`${databaseURL}/.json`);

const reqOptions = {
  hostname: urlObj.hostname,
  port: 443,
  path: urlObj.pathname + urlObj.search,
  method: 'PUT',
  headers: {
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(payload)
  }
};

const req = https.request(reqOptions, (res) => {
  let responseBody = '';
  res.on('data', (chunk) => responseBody += chunk);
  
  res.on('end', () => {
    if (res.statusCode === 200) {
      console.log("\n🎉 Success! Default database successfully seeded to Firebase Realtime Database.");
      console.log("You can now open the database in the Firebase Console to view the tables.");
    } else {
      console.error(`\n❌ Error Seeding Database (Status Code: ${res.statusCode}):`);
      try {
        console.error(JSON.stringify(JSON.parse(responseBody), null, 2));
      } catch (e) {
        console.error(responseBody);
      }
      console.error("\n💡 Hint: Make sure your Firebase Realtime Database Rules allow write access. In test mode, rules should look like:");
      console.error('{\n  "rules": {\n    ".read": "true",\n    ".write": "true"\n  }\n}');
    }
  });
});

req.on('error', (err) => {
  console.error("\n❌ Network Connection Error:", err.message);
});

req.write(payload);
req.end();

File: init_firebase.py
import os
import re
import json
import urllib.request
import hashlib
import secrets

CONFIG_PATH = 'firebase-config.js'
DB_PATH = 'db.json'

print("--------------------------------------------------")
print("[Fire] Firebase Database Initializer & Seeding Script (Python)")
print("--------------------------------------------------\n")

if not os.path.exists(CONFIG_PATH):
    print("[Error] firebase-config.js not found! Please create it first.")
    exit(1)

with open(CONFIG_PATH, 'r', encoding='utf-8') as f:
    config_content = f.read()

db_url_match = re.search(r'databaseURL\s*:\s*["\']([^"\']+)["\']', config_content)
if not db_url_match:
    print("[Error] Could not find databaseURL in firebase-config.js.")
    exit(1)

database_url = db_url_match.group(1)
if "YOUR_PROJECT_ID" in database_url:
    print("[Error] firebase-config.js is still using placeholder values.")
    exit(1)

if not os.path.exists(DB_PATH):
    print("[Error] Local db.json not found in this folder.")
    exit(1)

with open(DB_PATH, 'r', encoding='utf-8') as f:
    db_data = json.load(f)

print(f"Targeting Firebase Database: {database_url}")
print(f"Seeding settings, users ({len(db_data.get('users', []))}), allocations ({len(db_data.get('allocations', []))}), distributors ({len(db_data.get('distributors', []))})...")

# PBKDF2 SHA-256 Hashing helper
def hash_password(password: str, salt: str) -> str:
    return hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), salt.encode('utf-8'), 10000, 32).hex()

# Transform data to support slashes and match Firebase expectations
transformed_data = {}

# 1. Settings
transformed_data['settings'] = db_data.get('settings', {
    "writeUp": "Welcome to Adira Telecom Allocation Portal. Note: Confirmed allocations are strictly locked and credit will only be extended post proof verification.",
    "confirmationDeadline": "",
    "paymentDeadline": ""
})

# 2. Prices
transformed_data['prices'] = []
prices = db_data.get('prices', {})
if isinstance(prices, dict):
    for model_name, price in prices.items():
        transformed_data['prices'].append({
            'modelName': model_name,
            'price': float(price)
        })

# 3. Allocations
transformed_data['allocations'] = {}
allocations = db_data.get('allocations', [])
if isinstance(allocations, list):
    for item in allocations:
        if 'ID' in item:
            transformed_data['allocations'][item['ID']] = item

# 4. Distributors
transformed_data['distributors'] = {}
distributors = db_data.get('distributors', [])
if isinstance(distributors, list):
    for item in distributors:
        if 'AD Name' in item:
            # Replace invalid Firebase characters . # $ [ ] with _
            key = re.sub(r'[.#$\[\]]', '_', item['AD Name'])
            transformed_data['distributors'][key] = item

# 5. Users with Secure PBKDF2 Hashed Passwords
transformed_data['users'] = {}
users = db_data.get('users', [])
if isinstance(users, list):
    for item in users:
        if 'Username' in item:
            username = item['Username']
            plain_password = item.get('Password', '')
            
            # Generate a secure random salt (16 hex chars)
            salt = secrets.token_hex(8)
            hashed_password = hash_password(plain_password, salt)
            
            # Store credentials securely
            secure_user = item.copy()
            secure_user['Password'] = hashed_password
            secure_user['Salt'] = salt
            
            transformed_data['users'][username] = secure_user

payload = json.dumps(transformed_data, indent=2).encode('utf-8')

# PUT request using urllib
url = f"{database_url.rstrip('/')}/.json"
req = urllib.request.Request(
    url,
    data=payload,
    headers={
        'Content-Type': 'application/json',
        'Content-Length': len(payload)
    },
    method='PUT'
)

try:
    with urllib.request.urlopen(req) as response:
        status = response.status
        response_body = response.read().decode('utf-8')
        if status == 200:
            print("\n[Success] Default database successfully seeded with hashed credentials to Firebase Realtime Database.")
            print("You can now open the database in the Firebase Console to view the tables.")
        else:
            print(f"\n[Error] Seeding Database (Status Code: {status}):")
            print(response_body)
except urllib.error.HTTPError as e:
    print(f"\n[HTTP Error] (Status Code: {e.code}):")
    try:
        err_body = e.read().decode('utf-8')
        print(json.dumps(json.loads(err_body), indent=2))
    except Exception:
        print(e.reason)
    print("\nHint: Make sure your Firebase Realtime Database Rules allow write access. In test mode, rules should look like:")
    print('{\n  "rules": {\n    ".read": "true",\n    ".write": "true"\n  }\n}')
except Exception as e:
    print("\n[Network Error] Connection Error:", str(e))
